sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Relation.scala

```diff
@@ -17,17 +17,130 @@
 
 package org.apache.spark.sql.execution.datasources.v2
 
+import java.util.UUID
+
+import scala.collection.JavaConverters._
+import scala.collection.mutable
+
+import org.apache.spark.sql.{AnalysisException, SaveMode}
 import org.apache.spark.sql.catalyst.analysis.MultiInstanceRelation
-import org.apache.spark.sql.catalyst.expressions.AttributeReference
-import org.apache.spark.sql.catalyst.plans.logical.{LeafNode, Statistics}
-import org.apache.spark.sql.sources.v2.reader._
+import org.apache.spark.sql.catalyst.expressions.{AttributeReference, Expression}
+import org.apache.spark.sql.catalyst.plans.logical.{LeafNode, LogicalPlan, Statistics}
+import org.apache.spark.sql.execution.datasources.DataSourceStrategy
+import org.apache.spark.sql.sources.{DataSourceRegister, Filter}
+import org.apache.spark.sql.sources.v2.{DataSourceOptions, DataSourceV2, ReadSupport, ReadSupportWithSchema, WriteSupport}
+import org.apache.spark.sql.sources.v2.reader.{DataSourceReader, SupportsPushDownCatalystFilters, SupportsPushDownFilters, SupportsPushDownRequiredColumns, SupportsReportStatistics}
+import org.apache.spark.sql.sources.v2.writer.DataSourceWriter
+import org.apache.spark.sql.types.StructType
 
 case class DataSourceV2Relation(
-    output: Seq[AttributeReference],
-    reader: DataSourceReader)
-  extends LeafNode with MultiInstanceRelation with DataSourceReaderHolder {
+    source: DataSourceV2,
+    options: Map[String, String],
+    projection: Option[Seq[AttributeReference]] = None,
+    filters: Option[Seq[Expression]] = None,
+    userSchema: Option[StructType] = None) extends LeafNode with MultiInstanceRelation {
+
+  override def simpleString: String = {
+    s"DataSourceV2Relation(source=$sourceName, " +
+      s"schema=[${output.map(a => s"$a ${a.dataType.simpleString}").mkString(", ")}], " +
+      s"filters=[${pushedFilters.mkString(", ")}], options=$options)"
+  }
+
+  override lazy val schema: StructType = reader.readSchema()
+
+  override lazy val output: Seq[AttributeReference] = {
+    projection match {
+      case Some(attrs) =>
+        // use the projection attributes to avoid assigning new ids. fields that are not projected
+        // will be assigned new ids, which is okay because they are not projected.
+        val attrMap = attrs.map(a => a.name -> a).toMap
+        schema.map(f => attrMap.getOrElse(f.name,
+          AttributeReference(f.name, f.dataType, f.nullable, f.metadata)()))
+      case _ =>
+        schema.toAttributes
+    }
+  }
+
+  private lazy val v2Options: DataSourceOptions = {
+    // ensure path and table options are set correctly
+    val updatedOptions = new mutable.HashMap[String, String]
+    updatedOptions ++= options
+
+    new DataSourceOptions(options.asJava)
```

Comment:
This makes us create a new `DataSourceOptions` everytime `DataSourceV2Relation` gets transformed and copied. Can we just create `DataSourceOptions` at `DataFrameReader` and pass it to `DataSourceV2Relation`?

Reply:
As we've already discussed at length, I think it is a bad idea to create `DataSourceOptions` and pass it to the relation.

Reply:
Also, keep in mind that this is a lazy val. It is only referenced when creating a reader or writer

Reply:
We all agree that duplicating the logic of creating `DataSourceOptions` in many places is a bad idea. Currently there are 2 proposals:

1. Have a central place to take care the data source v2 resolution logic, including option creating. This is the approach of data source v1, i.e. the class `DataSource`.
2. Similar to proposal 1, but make `DataSourceV2Relation` the central place.

For now we don't know which one is better, it depends on how data source v2 evolves in the future. At this point of time, I think we should pick the simplest approach, which is passing the `DataSourceOptions` to `DataSourceV2Relation`. Then we just need a one-line change in `DataFrameReader`, and don't need to add `v2Options` here.

Reply:
I think you're missing that it's a bad idea to pass `DataSourceOptions` to the relation in the first place. As I've said, this class is specific to v2 readers and writers, and doesn't need to be leaked into logical plans.

To make this a bit more clear: If we pass the the options to the relation, then every time a new code path instantiates the relation, someone has to know where that central place to create options is. You're trading the problem for one only slightly better.

What if every relation required its own specific classes to pass the same information? That would be a mess. Rather than know what custom object to instantiate and where the helper for it is, it makes more sense to pass data to the relation. Then we can also use the info in match expressions.